### ISSUE TITLE
enable members to access search when master unavailable

### ISSUE THREAD

**Issue by Captainkirkdawson** [FreeUKGen/MyopicVicar#669]
We use a cookie to force members back to the master for work.
When master is unavailable they still get pushed there.
Need to determine how to get them back to round robin

**Comment by sean-wood** [comment]
waiting for information from Lemon

**Comment by lemon-ukgen** [comment]
So far we have the following detection for requests that go to the master:
```
acl reg2_master path_beg -f /usr/local/etc/haproxy/freereg2-masterscripts # /refinery-prefixed pats
acl reg2_master req.cook_cnt(Administrator) gt 0 # cookie set by Devise (?) during login
```

We also detect search queries:

```
acl reg2_search path_beg /search_queries/
```

Previously if any of the first two were true we'd send the request to only the master

```
use_backend reg2_master if reg2_master reg2_live
```

I've added a new backend which has the master server as the usual server and the other servers set as backups. This is used when all three conditions are true:

```
use_backend reg2_master_with_fallback if reg2_master reg2_live reg2_search
```
There might be a strange interaction between the sticky cookie we set when sending a search request to one of the servers (to make sure further searches continue to that server) and the active/backup composition of this `reg2_master_with_fallback` backend: In the case of the master being unavailable this cookie will guide requests to a non-master server (which is the purpose of this ticket) but what happens when the master becomes available again. I think HAProxy will continue to send the request to the backup server that was previously chosen. HAProxy doc says about backup servers: [Requests coming
with a persistence cookie referencing the server will always be served
though.](https://cbonte.github.io/haproxy-dconv/configuration-1.5.html#5.2-backup)

It may not be worth worrying too much about that.

**Comment by Captainkirkdawson** [comment]
There is an unanticipated impact of this change. 

It would appear that members who login to the site are then unable to complete a search. They can enter the search criteria. The search will be performed but when the system goes to show the results it cannot find the search_query.id and the user gets the message 
We encountered a problem completing your request. Please resubmit. If this situation continues please let us know through the Contact Us link at the foot of this page 
This message is generated by the search_query show action when it cannot link to the search query. We catch and generate the message rather than throw a mongoid crash.

I believe that we are running the search on a different server from the one where the search query was generated.

For example. I clear my cookies. I access the freereg. I get assigned to brazza. I do a search OK. I log in and the admin cookie is set, the sessionserver cookie is still set for brazza. I can do all my member actions BUT if I do a search I get the error message.

A search of the brazza log shows the message
Started GET "/search_queries/569feba8f493fd1b4d0003c9" for 5.9.125.163 at 2016-01-20 20:18:48 +0000
Processing by SearchQueriesController#show as HTML
  Parameters: {"id"=>"569feba8f493fd1b4d0003c9"}
FREEREG:SEARCH_ERROR:search query no longer present

In other words the search query has been stored on private writeable collection on colobus but the search was run on brazza which cannot find the search query.

This message has been generated on brazza 92 times today. The first time was 00:32 GMT

I suggest that if we cannot quickly see the cause of this behaviour we revert to the previous HaProxy code.

**Comment by lemon-ukgen** [comment]
[ moved this to GH ticket for reference, but I'll attend to the other matter first ]

On 20/01/2016 23:02 ICT Kirk Dawson wrote:
> AS further question.
> 
> I notice that colobus is currently shown as down for some apps but
> not all this I presume is due to the BMD update which is running,
> 
> The previous status used to be maintenance did it not? The down time
> is 12 hrs so expect that this was part of the change.
> 
> This will activate the new arrangement of sending all searches to a
> backup regardless of the server cookie?
> 
> The disable we use on our update will also set to down? and redirect
> all searches regardless of sticky away from the master? or?
> 

The `searchctl` method of disabling search on the master is only used by FreeREG2 (as far as I know) and places colobus in to maintenance in all `*_search` backends.

The healthcheck FreeBMD contains a test for a database build in progress. This would mark colobus down for just the `bmd_search` backend.

This shouldn't in theory affect FreeREG2 *but* in error FreeREG2 is using FreeBMD's healthcheck target (`/cgi/service.pl`).

We should create FreeREG2's own check - a lightweight request that verifies the server is functional - this can just be `/` but that's likely too burdensome.

Without a healthcheck HAProxy will just test port 80 is open which is less useful.

For now I'll make FreeREG2 use FreeREG1's healthcheck which is a fudge but it will at least always indicate the server is up regardless of what FreeBMD is using, since the FreeREG1 database will never be rebuilt again and the VirtualHost is present on all servers still.

**Comment by lemon-ukgen** [comment]
On 21/01/2016 04:39 ICT Kirk Dawson wrote:

> [....]
> 
> In other words the search query has been stored on private writeable
> collection on colobus but the search was run on brazza which cannot
> find the search query.

Does this sounds like that we're not properly identifying the start of
the search? I mean: does

```
acl reg2_search path_beg /search_queries/
```

properly capture any request that creates or uses this private data?

It sounds like it doesn't, otherwise the private data in this example
would have been created on brazza, not colobus.

> 
> This message has been generated on brazza 92 times today. The first
> time was 00:32 GMT
> 
> I suggest that if we cannot quickly see the cause of this behaviour
> we revert to the previous HaProxy code.

Done.

**Comment by lemon-ukgen** [comment]
On 21/01/2016 04:39 ICT Kirk Dawson wrote:

> [....]
> 
> In other words the search query has been stored on private writeable
> collection on colobus but the search was run on brazza which cannot
> find the search query.

Does this sounds like that we're not properly identifying the start of
the search? I mean: does

```
acl reg2_search path_beg /search_queries/
```

properly capture any request that creates or uses this private data?

It sounds like it doesn't, otherwise the private data in this example
would have been created on brazza, not colobus.

> 
> This message has been generated on brazza 92 times today. The first
> time was 00:32 GMT
> 
> I suggest that if we cannot quickly see the cause of this behaviour
> we revert to the previous HaProxy code.

Done.

**Comment by Captainkirkdawson** [comment]
Thanks for the reversion. Since I am now getting ready to leave on vacation, I leave it to Ben/Doug to decide on next move. 

Wrt your comment "properly capture any request that creates or uses this private data? It sounds like it doesn't, otherwise the private data in this example would have been created on brazza, not colobus"

I am seeing the odd occasional other issue that makes me perhaps share this concern. See issue #707 where a rake task is running on a slave which can only be initiated by a member who is supposed to only be on the master. This happens very very occasionally. So how did the rake task get to the slave? presumably only if the request to initiate the task somehow had gone off to the slave first but why would this happen? 
Have fun guys!!

**Comment by sean-wood** [comment]
27th Jan - moved to product backlog as changes were backed out due to other issues, this will need to be re-visited if it occurs again

**Comment by lemon-ukgen** [comment]
Looking at the HAProxy logs for yesterday's failover I notice that search `POST`s go to `/search_queries` not `/search_queries/` so won't be served by the `reg2_search` backend, which is used by requests that have the trailing slash.

This doesn't matter usually, all the backends use the same session affinity cookie so regardless of which backend you go to, the cookie will be honoured and you'll hit the server with the search data.

However, I can see it might have a bad interaction with the `reg2_master_with_fallback` rules. `GET`s for logged in users will always use the right server -- colobus if alive, others if not -- , but `POST`s may not, and the request goes to wherever the affinity cookie says. This matches @Captainkirkdawson's observations above.

It's a little unclear, perhaps the best thing to do is to update the regex and test again. Let me know if you'd like to do this.

**Comment by Captainkirkdawson** [comment]
Will not be relevant under new server configuration
